go/types/simplify.go

```diff
@@ -191,68 +167,51 @@ func simplifyMaps(ts typeset, merge bool) *Type {
 		valTypes.Add(desc.ElemTypes[1])
 	}
 	return MakeMapType(
-		makeSimplifiedTypeImpl(keyTypes, merge),
-		makeSimplifiedTypeImpl(valTypes, merge))
+		makeSimplifiedTypeImpl(keyTypes),
+		makeSimplifiedTypeImpl(valTypes))
 }
 
-func simplifyStructs(expectedName string, ts typeset, merge bool) *Type {
-	if merge {
-		return mergeStructs(expectedName, ts)
+func simplifyStructs(expectedName string, ts typeset) *Type {
+	// We gather all the fields/types into commonFields. If the number of
+	// times a field name is present is less that then number of types we
+	// are simplifying then the field must be optional.
+	// If we see an optional field we do not increment the count for it and
+	// it will be treated as optional in the end.
+	type fieldTypeInfo struct {
+		count   int
+		typeset typeset
 	}
-	commonFields := map[string]typeset{}
+	commonFields := map[string]fieldTypeInfo{}
 
-	first := true
 	for t := range ts {
 		d.Chk.True(StructKind == t.Kind())
 		desc := t.Desc.(StructDesc)
 		d.Chk.True(expectedName == desc.Name)
-		if first {
-			for _, f := range desc.fields {
-				ts := typeset{}
-				ts.Add(f.t)
-				commonFields[f.name] = ts
-			}
-			first = false
-		} else {
-			for n, ts := range commonFields {
-				t := desc.Field(n)
-				if t != nil {
-					ts.Add(t)
-				} else {
-					delete(commonFields, n)
+		desc.IterFields(func(name string, t *Type, optional bool) {
+			fti, ok := commonFields[name]
+			if !ok {
+				fti = fieldTypeInfo{
+					count:   0,
+					typeset: typeset{},
 				}
 			}
-		}
-	}
-
-	fm := make(FieldMap, len(ts))
-	for n, ts := range commonFields {
-		fm[n] = makeSimplifiedTypeImpl(ts, false)
-	}
-
-	return MakeStructTypeFromFields(expectedName, fm)
-}
-
-func mergeStructs(expectedName string, ts typeset) *Type {
-	unionFields := map[string]typeset{}
-	for t := range ts {
-		d.Chk.True(StructKind == t.Kind())
-		desc := t.Desc.(StructDesc)
-		d.Chk.True(expectedName == desc.Name)
-		for _, f := range desc.fields {
-			ts, ok := unionFields[f.name]
-			if !ok {
-				ts = typeset{}
+			fti.typeset.Add(t)
+			if !optional {
+				fti.count++
 			}
-			ts.Add(f.t)
-			unionFields[f.name] = ts
-		}
+			commonFields[name] = fti
```

Comment:
You don't need to add this back to the map in the case where it was already there